opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -844,6 +868,54 @@ namespace Opm {
 
 
 
+    template<typename TypeTag>
+    void
+    BlackoilWellModel<TypeTag>::
+    balanceNetwork(DeferredLogger& deferred_logger) {
+        const double dt = this->ebosSimulator_.timeStepSize();
+        // updatePerforationIntensiveQuantities();
+        // TODO: should we also have the group and network backed-up here?
+        auto& well_state = this->wellState();
+        // B_avg is updated in the beginTimeStep();
+        const int max_iter = 100;
+        bool converged = false;
+        size_t iter = 0;
+        bool changed_well_group = false;
+        do {
+            changed_well_group = updateWellControlsAndNetwork(deferred_logger, true);
+            assembleImpl(dt, deferred_logger);
+            converged = this->getWellConvergence(this->B_avg_, true).converged() && !changed_well_group;
+#if EXTRA_NETWORK_OUTPUT
+            std::cout << " balanceNetwork iteration " << iter << " converged ?";
+            if (converged) {
+                std::cout << " YES ";
+            } else {
+                std::cout << " NO ";
+            }
+#endif
+            if (converged) {
+                break;
+            }
+            ++iter;
+            for (auto& well : this->well_container_) {
+                well->solveEqAndUpdateWellState(well_state, deferred_logger);
+            }
+            this->initPrimaryVariablesEvaluation();
+        } while (iter < max_iter);
```

Comment:
Maybe add a check to see if we converged after the loop, and print a warning "Network balancing failed" or similar? Or even better: return a bool (true if converged), and let the calling function deal with it.